pmd-ui/src/main/java/net/sourceforge/pmd/util/fxdesigner/MainDesignerController.java

```diff
@@ -0,0 +1,475 @@
+import javafx.scene.control.SeparatorMenuItem;
+import javafx.scene.control.SplitPane;
+import javafx.scene.control.Tab;
+import javafx.scene.control.TabPane;
+import javafx.scene.control.TextArea;
+import javafx.scene.control.ToggleButton;
+import javafx.scene.control.Tooltip;
+import javafx.stage.FileChooser;
+import javafx.stage.Modality;
+import javafx.stage.Stage;
+import javafx.util.Duration;
+import javafx.util.StringConverter;
+
+
+/**
+ * Main controller of the app. Mediator for subdivisions of the UI.
+ *
+ * @author Clément Fournier
+ * @see NodeInfoPanelController
+ * @see SourceEditorController
+ * @see EventLogController
+ * @see XPathPanelController
+ * @since 6.0.0
+ */
+public class MainDesignerController implements Initializable, SettingsOwner {
+
+    /**
+     * Callback to the owner.
+     */
+    private final DesignerRoot designerRoot;
+
+    /* Menu bar */
+    @FXML
+    private MenuItem openFileMenuItem;
+    @FXML
+    private MenuItem licenseMenuItem;
+    @FXML
+    private Menu openRecentMenu;
+    @FXML
+    private MenuItem exportToTestCodeMenuItem;
+    @FXML
+    private MenuItem exportXPathMenuItem;
+    @FXML
+    private Menu exportMenu;
+    @FXML
+    private Menu fileMenu;
+    /* Center toolbar */
+    @FXML
+    private Button refreshASTButton;
+    @FXML
+    private ChoiceBox<LanguageVersion> languageChoiceBox;
+    @FXML
+    private ChoiceBox<String> xpathVersionChoiceBox;
+    @FXML
+    private ToggleButton bottomTabsToggle;
+    /* Bottom panel */
+    @FXML
+    private TabPane bottomTabPane;
+    @FXML
+    private Tab eventLogTab;
+    @FXML
+    private Tab xpathEditorTab;
+    @FXML
+    private SplitPane mainHorizontalSplitPane;
+    /* Children */
+    @FXML
+    private NodeInfoPanelController nodeInfoPanelController;
+    @FXML
+    private XPathPanelController xpathPanelController;
+    @FXML
+    private SourceEditorController sourceEditorController;
+    @FXML
+    private EventLogController eventLogPanelController;
+    private Stack<File> recentFiles = new LimitedSizeStack<>(5);
+
+
+    public MainDesignerController(DesignerRoot owner) {
+        this.designerRoot = owner;
+    }
+
+
+    @Override
+    public void initialize(URL location, ResourceBundle resources) {
+
+        try {
+            XMLSettingsLoader loader = new XMLSettingsLoader(DesignerUtil.getSettingsFile());
+            loadSettings(loader.getSettings());
+        } catch (IOException ioe) {
+            // no big deal
+        }
+
+
+        initializeLanguageVersionMenu();
+        initializeViewAnimation();
+
+        xpathPanelController.initialiseVersionChoiceBox(xpathVersionChoiceBox);
+
+        sourceEditorController.languageVersionProperty().bind(languageChoiceBox.getSelectionModel().selectedItemProperty());
+        xpathPanelController.xpathVersionProperty().bind(xpathVersionChoiceBox.getSelectionModel().selectedItemProperty());
+
+        refreshASTButton.setOnAction(e -> onRefreshASTClicked());
```

Comment:
You can also define these (action) events in fxml, e.g.

`<Button fx:id="refreshASTButton" ... onAction="#onRefreshASTClicked">`

The method can stay private, but you'll need to annotate it with "@FXML".
The same with them menu items.

That way you probably don't need the instances of the buttons, menuitems in the controller and you can get rid of the `fx:id` attributes in fxml.

This should work theoretically. I didn't use a custom ControllerFactory on the FXMLLoader yet, but I guess, it is the same as using the default implementation - the fxml as a specific instance of a controller attached, so the connection between fxml and controller should work...

Reply:
true, but would couple further the fxml with the code... not sure if saving 1 id + 1 LoC is worth it.... but this is my first time using JFX, so don't take me too seriously

Reply:
Hum, I kind of agree with Juan in that, ideally, the UI should be as dumb as possible. Plus, if we want to fiddle with this button someday it will be convenient to have it already available in the controller I think. So since it's already done, I'd say it's better to keep it that way